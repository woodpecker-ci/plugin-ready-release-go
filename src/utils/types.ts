import type { ExecFunction } from 'shelljs';
import type { Forge, PullRequest } from '../forges/forge';
import type { Config } from './config';
import type { SimpleGit } from 'simple-git';

export type PromiseOrValue<T> = Promise<T> | T;

export type Exec = ExecFunction;

export type CommandContext = {
  exec: Exec;
  config: Config;
  forge: Forge;
  git: SimpleGit;
  latestVersion: string;
  useVersionPrefixV: boolean;
  nextVersion: string;
  pullRequestBranch: string;
  shouldBeRC: boolean;
  changes: Change[];
};

export type Change = {
  title: string;
  author: string;
  commitHash: string;
  pullRequestNumber?: number;
  labels: string[];
};

export type HookContext = {
  exec: Exec;
  latestVersion?: string;
  nextVersion?: string;
  changes?: Change[];
};

export type UserConfig = Partial<{
  /**
   * Get the next version to release. By default pull-request labels will be used.
   */
  getNextVersion: (ctx: HookContext) => PromiseOrValue<string>;

  /**
   * Get the release description.
   *
   * Used as pull-request and release description.
   * By default it's the changelog part of the current / next version.
   */
  getReleaseDescription: (ctx: HookContext) => PromiseOrValue<string>;

  /**
   * Run before the release pull request is created.
   *
   * Could be used to update changelog, version, etc
   */
  beforePrepare: (ctx: HookContext) => PromiseOrValue<boolean | void>;

  /**
   * Run after the release pull request was created.
   *
   * Could be used to notify maintainers about review, etc
   */
  afterPrepare: (ctx: HookContext) => PromiseOrValue<boolean | void>;

  /**
   * Run before the release is created.
   *
   * Could be used to create release notes, etc
   */
  beforeRelease: (ctx: HookContext) => PromiseOrValue<boolean | void>;

  /**
   * Run after the release was created.
   *
   * Could be used to notify maintainers about release, publish release artifacts, etc
   */
  afterRelease: (ctx: HookContext) => PromiseOrValue<boolean | void>;

  changeTypes: {
    title: string;
    labels: string[];
    bump: 'major' | 'minor' | 'patch';
    default?: boolean;
    weight?: number;
  }[];

  /**
   * Skip commits associated with pull-requests with these labels
   * @default ["skip-release", "skip-changelog", "regression"]
   */
  skipLabels: string[];

  /**
   * Skip commits that are not associated with a pull-request
   * @default true
   */
  skipCommitsWithoutPullRequest: boolean;

  /**
   * Use a 'v' prefix for the version
   * @default true if the previous version already has a v-prefix
   */
  useVersionPrefixV: boolean;

  /**
   * Comment on pull-requests that are included in the release
   * @default true
   */
  commentOnReleasedPullRequests: boolean;

  /**
   * Run semver sort over tags before detecting latest tag
   * @default true
   */
  sortTags: boolean;
}>;

export const defineConfig = (config: UserConfig) => config;
