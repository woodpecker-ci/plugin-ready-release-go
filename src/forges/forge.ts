export type Comment = {
  id: string;
  author: string;
  body: string;
  createdAt: string;
};

export type PullRequest = {
  number: number;
  title: string;
  author: string;
  description: string;
  labels: string[];
};

export abstract class Forge {
  abstract createOrUpdatePullRequest(options: {
    draft: boolean;
    title: string;
    description: string;
    owner: string;
    repo: string;
    sourceBranch: string;
    targetBranch: string;
  }): Promise<{
    pullRequestLink: string;
  }>;

  abstract createRelease(options: {
    owner: string;
    repo: string;
    tag: string;
    name: string;
    description: string;
    prerelease?: boolean;
  }): Promise<{
    releaseLink: string;
  }>;

  abstract addCommentToPullRequest(options: {
    owner: string;
    repo: string;
    pullRequestNumber: number;
    comment: string;
  }): Promise<void>;

  abstract getGitCredentials(): Promise<{
    email: string;
    username: string;
    password: string;
  }>;

  abstract getPullRequestFromCommit(options: {
    owner: string;
    repo: string;
    commitHash: string;
  }): Promise<PullRequest | undefined>;

  abstract getPullRequest(options: {
    owner: string;
    repo: string;
    sourceBranch: string;
    targetBranch: string;
  }): Promise<PullRequest | undefined>;

  abstract getRepoUrl(owner: string, repo: string): string;
  abstract getCommitUrl(owner: string, repo: string, commitHash: string): string;
  abstract getIssueUrl(owner: string, repo: string, issueNumber: string | number): string;
  abstract getPullRequestUrl(owner: string, repo: string, pullRequestNumber: string | number): string;
  abstract getReleaseUrl(owner: string, repo: string, release: string): string;
}
