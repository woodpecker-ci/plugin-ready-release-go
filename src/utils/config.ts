import path from 'path';
import { UserConfig } from './types';
import { promises as fs } from 'fs';

const ciConfig = {
  configFile: process.env.PLUGIN_CONFIG_FILE,
  isCI: process.env.CI === 'woodpecker',
  eventType: process.env.CI_PIPELINE_EVENT,
  releaseBranch: process.env.PLUGIN_RELEASE_BRANCH || process.env.CI_REPO_DEFAULT_BRANCH || 'main',
  commitMessage: process.env.CI_COMMIT_MESSAGE,
  forgeType: process.env.CI_FORGE_TYPE,
  githubToken: process.env.PLUGIN_GITHUB_TOKEN,
  gitEmail: process.env.PLUGIN_GIT_EMAIL,
  repoOwner: process.env.CI_REPO_OWNER,
  repoName: process.env.CI_REPO_NAME,
  pullRequestBranchPrefix: process.env.PLUGIN_PULL_REQUEST_BRANCH_PREFIX || 'next-release/',
  releasePrefix: '🎉 Release',
  debug: process.env.PLUGIN_DEBUG === 'true',
};

export type Config = { user: UserConfig; ci: typeof ciConfig };

export const defaultUserConfig: UserConfig = {
  changeTypes: [
    {
      title: '💥 Breaking changes',
      labels: ['breaking'],
      bump: 'major',
      weight: 3,
    },
    {
      title: '🔒 Security',
      labels: ['security'],
      bump: 'patch',
      weight: 2,
    },
    {
      title: '✨ Features',
      labels: ['feature', 'feature 🚀️'],
      bump: 'minor',
      weight: 1,
    },
    {
      title: '📈 Enhancement',
      labels: ['enhancement', 'refactor', 'enhancement 👆️'],
      bump: 'minor',
    },
    {
      title: '🐛 Bug Fixes',
      labels: ['bug', 'bug 🐛️'],
      bump: 'patch',
    },
    {
      title: '📚 Documentation',
      labels: ['docs', 'documentation', 'documentation 📖️'],
      bump: 'patch',
    },
    {
      title: 'Misc',
      labels: ['misc', 'chore 🧰'],
      bump: 'patch',
      default: true,
    },
    {
      title: '⚙️ Dependency',
      labels: ['dependency', 'dependencies'],
      bump: 'patch',
      default: true,
    },
  ],
  skipLabels: ['skip-release', 'skip-changelog', 'regression'],
  skipCommitsWithoutPullRequest: true,
  commentOnReleasedPullRequests: true,
};

export async function getConfig(): Promise<Config> {
  const userConfig: UserConfig = {};

  const configFilePath = ciConfig.configFile || path.join(process.cwd(), 'release-config.ts');
  if (
    await fs
      .stat(configFilePath)
      .then(() => true)
      .catch(() => false)
  ) {
    console.log('# Loading config from', configFilePath, '...');
    const _userConfig = await import(configFilePath);
    Object.assign(userConfig, _userConfig.default);
    console.log('# Loaded config from', configFilePath);
  }

  return {
    user: { ...defaultUserConfig, ...userConfig },
    ci: ciConfig,
  };
}
