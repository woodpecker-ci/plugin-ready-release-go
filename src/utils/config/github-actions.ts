import { CIConfig } from '../types';

export function getGithubActionsConfig(): Partial<CIConfig> {
  const eventType = process.env.GITHUB_EVENT_NAME; // TODO: map to woodpecker event types

  const githubToken = process.env.INPUT_GITHUB_TOKEN || process.env.GITHUB_TOKEN;
  if (!githubToken) {
    throw new Error('Please provide a GitHub token');
  }

  return {
    configFile: process.env.INPUT_CONFIG_FILE,
    eventType,
    releaseBranch: process.env.INPUT_RELEASE_BRANCH || process.env.CI_REPO_DEFAULT_BRANCH,
    commitMessage: process.env.CI_COMMIT_MESSAGE,
    forgeType: 'github',
    githubToken,
    gitEmail: process.env.INPUT_GIT_EMAIL,
    repoOwner: process.env.GITHUB_REPOSITORY_OWNER,
    repoName: process.env.GITHUB_REPOSITORY?.split('/')?.[1],
    pullRequestBranchPrefix: process.env.INPUT_PULL_REQUEST_BRANCH_PREFIX,
    debug: process.env.INPUT_DEBUG === 'true',
  };
}
