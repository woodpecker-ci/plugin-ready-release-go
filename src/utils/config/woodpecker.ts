import { CIConfig } from '../types';

export function getWoodpeckerConfig(): Partial<CIConfig> {
  return {
    configFile: process.env.PLUGIN_CONFIG_FILE,
    eventType: process.env.CI_PIPELINE_EVENT,
    releaseBranch: process.env.PLUGIN_RELEASE_BRANCH || process.env.CI_REPO_DEFAULT_BRANCH,
    commitMessage: process.env.CI_COMMIT_MESSAGE,
    forgeType: process.env.CI_FORGE_TYPE,
    githubToken: process.env.PLUGIN_GITHUB_TOKEN,
    gitEmail: process.env.PLUGIN_GIT_EMAIL,
    repoOwner: process.env.CI_REPO_OWNER,
    repoName: process.env.CI_REPO_NAME,
    pullRequestBranchPrefix: process.env.PLUGIN_PULL_REQUEST_BRANCH_PREFIX,
    debug: process.env.PLUGIN_DEBUG === 'true',
  };
}
