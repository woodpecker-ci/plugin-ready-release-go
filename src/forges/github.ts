import { Comment, Forge, PullRequest } from './forge';
import { Octokit } from '@octokit/rest';

export class GithubForge extends Forge {
  accessToken: string;
  email: string;
  octokit: Octokit;

  constructor(accessToken: string, email: string) {
    super();
    this.email = email;
    this.accessToken = accessToken;
    this.octokit = new Octokit({
      auth: accessToken,
    });
  }

  async createOrUpdatePullRequest(options: {
    draft: boolean;
    title: string;
    description: string;
    owner: string;
    repo: string;
    sourceBranch: string;
    targetBranch: string;
  }): Promise<{ pullRequestLink: string }> {
    const pullRequests = await this.octokit.pulls.list({
      owner: options.owner,
      repo: options.repo,
      head: `${options.owner}:${options.sourceBranch}`,
    });

    if (pullRequests.data.length > 0) {
      const pr = await this.octokit.pulls.update({
        owner: options.owner,
        repo: options.repo,
        pull_number: pullRequests.data[0].number,
        title: options.title,
        draft: options.draft,
        head: options.sourceBranch,
        base: options.targetBranch,
        body: options.description,
      });

      return { pullRequestLink: pr.data._links.html.href };
    }

    const pr = await this.octokit.pulls.create({
      owner: options.owner,
      repo: options.repo,
      title: options.title,
      draft: options.draft,
      head: options.sourceBranch,
      base: options.targetBranch,
      body: options.description,
    });

    return { pullRequestLink: pr.data._links.html.href };
  }

  async createRelease(options: {
    owner: string;
    repo: string;
    tag: string;
    name: string;
    description: string;
    prerelease?: boolean;
  }): Promise<{ releaseLink: string }> {
    const release = await this.octokit.repos.createRelease({
      owner: options.owner,
      repo: options.repo,
      tag_name: options.tag,
      name: options.name,
      body: options.description,
      prerelease: options.prerelease,
    });

    return { releaseLink: release.data.html_url };
  }

  async getGitCredentials(): Promise<{
    email: string;
    username: string;
    password: string;
  }> {
    return {
      email: this.email,
      username: 'oauth',
      password: this.accessToken,
    };
  }

  async getPullRequestFromCommit(options: {
    owner: string;
    repo: string;
    commitHash: string;
  }): Promise<PullRequest | undefined> {
    const pr = await this.octokit.repos.listPullRequestsAssociatedWithCommit({
      owner: options.owner,
      repo: options.repo,
      commit_sha: options.commitHash,
    });

    if (pr.data.length === 0) {
      return undefined;
    }

    return {
      number: pr.data[0].number,
      title: pr.data[0].title,
      description: pr.data[0].body || '',
      author: pr.data[0].user?.login || '',
      labels: pr.data[0].labels.map((label) => label.name),
    };
  }

  async getPullRequest(options: {
    owner: string;
    repo: string;
    sourceBranch: string;
    targetBranch: string;
  }): Promise<PullRequest | undefined> {
    const pullRequests = await this.octokit.pulls.list({
      owner: options.owner,
      repo: options.repo,
      head: `${options.owner}:${options.sourceBranch}`,
    });

    if (pullRequests.data.length > 1) {
      throw new Error('Found more than one pull request to release. Please close all but one.');
    }

    if (pullRequests.data.length === 0) {
      return undefined;
    }

    const pr = pullRequests.data[0];

    return {
      number: pr.number,
      title: pr.title,
      description: pr.body || '',
      author: pr.user?.login || '',
      labels: pr.labels.map((label) => label.name),
    };
  }

  async addCommentToPullRequest(options: {
    owner: string;
    repo: string;
    pullRequestNumber: number;
    comment: string;
  }): Promise<void> {
    await this.octokit.issues.createComment({
      owner: options.owner,
      repo: options.repo,
      issue_number: options.pullRequestNumber,
      body: options.comment,
    });
  }

  getRepoUrl(owner: string, repo: string): string {
    return `https://github.com/${owner}/${repo}`;
  }

  getCommitUrl(owner: string, repo: string, commitHash: string): string {
    return `https://github.com/${owner}/${repo}/commit/${commitHash}`;
  }

  getIssueUrl(owner: string, repo: string, issueNumber: string | number): string {
    return `https://github.com/${owner}/${repo}/issues/${issueNumber}`;
  }

  getPullRequestUrl(owner: string, repo: string, pullRequestNumber: string | number): string {
    return `https://github.com/${owner}/${repo}/pull/${pullRequestNumber}`;
  }

  getReleaseUrl(owner: string, repo: string, release: string): string {
    return `https://github.com/${owner}/${repo}/releases/tag/${release}`;
  }

  async getPullRequestComments(owner: string, repo: string, pullRequestNumber: number): Promise<Comment[]> {
    const comments = await this.octokit.paginate(this.octokit.issues.listComments, {
      owner,
      repo,
      issue_number: pullRequestNumber,
    });

    return comments.map((comment) => ({
      id: comment.id.toString(),
      body: comment.body!,
      author: comment.user?.login!,
      createdAt: comment.created_at,
    }));
  }
}
