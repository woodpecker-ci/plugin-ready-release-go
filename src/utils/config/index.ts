import path from 'path';
import { UserConfig } from '../types';
import { promises as fs } from 'fs';
import type { CIConfig, Config } from '../types';
import { getWoodpeckerConfig } from './woodpecker';
import { getGithubActionsConfig } from './github-actions';

export const defaultUserConfig: UserConfig = {
  changeTypes: [
    {
      title: '💥 Breaking changes',
      labels: ['breaking'],
      bump: 'major',
      weight: 3,
    },
    {
      title: '🔒 Security',
      labels: ['security'],
      bump: 'patch',
      weight: 2,
    },
    {
      title: '✨ Features',
      labels: ['feature', 'feature 🚀️'],
      bump: 'minor',
      weight: 1,
    },
    {
      title: '📈 Enhancement',
      labels: ['enhancement', 'refactor', 'enhancement 👆️'],
      bump: 'minor',
    },
    {
      title: '🐛 Bug Fixes',
      labels: ['bug', 'bug 🐛️'],
      bump: 'patch',
    },
    {
      title: '📚 Documentation',
      labels: ['docs', 'documentation', 'documentation 📖️'],
      bump: 'patch',
    },
    {
      title: 'Misc',
      labels: ['misc', 'chore 🧰'],
      bump: 'patch',
      default: true,
    },
  ],
  skipLabels: ['skip-release', 'skip-changelog', 'regression'],
  skipCommitsWithoutPullRequest: true,
  commentOnReleasedPullRequests: true,
};

const defaultCIConfig: CIConfig = {
  eventType: 'manual',
  releaseBranch: 'main',
  commitMessage: '',
  forgeType: 'github',
  githubToken: '',
  gitEmail: '',
  repoOwner: '',
  repoName: '',
  pullRequestBranchPrefix: 'next-release/',
  releasePrefix: '🎉 Release',
  debug: false,
};

export async function getConfig(): Promise<Config> {
  const userConfig: UserConfig = {};
  let ciConfig: CIConfig = defaultCIConfig;

  if (process.env.CI === 'woodpecker') {
    console.log('# CI: Woodpecker');
    ciConfig = { ...defaultCIConfig, ...getWoodpeckerConfig() };
  } else if (process.env.GITHUB_ACTIONS === 'true') {
    console.log('# CI: GitHub Actions');
    ciConfig = { ...defaultCIConfig, ...getGithubActionsConfig() };
  } else {
    throw new Error('Unsupported CI type');
  }

  const configFilePath = ciConfig.configFile || path.join(process.cwd(), 'release-config.ts');
  if (
    await fs
      .stat(configFilePath)
      .then(() => true)
      .catch(() => false)
  ) {
    console.log('# Loading config from', configFilePath, '...');
    const _userConfig = await import(configFilePath);
    Object.assign(userConfig, _userConfig.default);
    console.log('# Loaded config from', configFilePath);
  }

  return {
    user: { ...defaultUserConfig, ...userConfig },
    ci: ciConfig,
  };
}
