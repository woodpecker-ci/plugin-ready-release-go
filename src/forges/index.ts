import { Config } from '../utils/config';
import { Forge } from './forge';
import { GithubForge } from './github';

export async function getForge(config: Config): Promise<Forge> {
  if (config.ci.forgeType === 'github') {
    if (!config.ci.githubToken) {
      throw new Error('Please provide a Github token');
    }

    if (!config.ci.gitEmail) {
      throw new Error('Please provide a Git email');
    }

    return new GithubForge(config.ci.githubToken, config.ci.gitEmail);
  }

  throw new Error('Forge type not supported: ' + config.ci.forgeType);
}
