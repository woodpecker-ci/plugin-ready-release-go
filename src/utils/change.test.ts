import { describe, it, expect, vi, beforeAll } from 'vitest';
import {
  extractVersionFromCommitMessage,
  getChangeLogSection,
  getNextVersionFromLabels,
  updateChangelogSection,
} from './change';
import { Change } from './types';
import { Config, defaultUserConfig } from './config';
import { GithubForge } from '../forges/github';
import { promises as fs } from 'fs';
import path from 'path';

const changes: Change[] = [
  {
    commitHash: '123',
    author: 'John Doe',
    labels: ['bug', 'ui'],
    title: 'Fix random UI bug',
    pullRequestNumber: 1337,
  },
  {
    commitHash: '456',
    author: 'Alice Wonderland',
    labels: ['feature'],
    title: 'Add new feature',
    pullRequestNumber: 1338,
  },
  {
    commitHash: '789',
    author: 'John Doe',
    labels: ['breaking'],
    title: 'Break everything',
    pullRequestNumber: 1339,
  },
  {
    commitHash: 'abc',
    author: 'Alice Wonderland',
    labels: ['bug'],
    title: 'Fix another bug',
  },
  {
    commitHash: 'def',
    author: 'Alice Wonderland',
    labels: [],
    title: 'Update README',
  },
  {
    commitHash: 'def',
    author: 'Alice Wonderland',
    labels: ['enhancement'],
    title: 'Update README',
  },
];

const changesWithMajor = changes;
const changesWithMinor = changes.filter((change) => !change.labels.includes('breaking'));
const changesWithPatch = changes.filter(
  (change) =>
    !change.labels.includes('breaking') && !change.labels.includes('feature') && !change.labels.includes('enhancement'),
);

const config: Config = {
  ci: {
    releaseBranch: 'main',
    commitMessage: 'chore(release): 1.0.0 [skip ci]',
    configFile: 'ready-go-release.config.ts',
    eventType: 'push',
    gitEmail: 'ci@woodpecker-ci.org',
    repoOwner: 'woodpecker-ci',
    repoName: 'woodpecker',
    forgeType: 'github',
    forgeToken: '123',
    forgeURL: '',
    pullRequestBranchPrefix: 'next-release/',
    isCI: true,
    releasePrefix: 'ðŸŽ‰ Release',
    debug: false,
  },
  user: defaultUserConfig,
};

describe('change', () => {
  beforeAll(() => {
    const date = new Date(2000, 1, 1, 13);
    vi.setSystemTime(date);
  });

  it('should get the correct major bump', () => {
    const nextVersion = getNextVersionFromLabels('1.0.0', config.user, changesWithMajor, false);
    expect(nextVersion).toBe('2.0.0');
  });

  it('should get the correct minor bump', () => {
    const nextVersion = getNextVersionFromLabels('1.0.0', config.user, changesWithMinor, false);

    expect(nextVersion).toBe('1.1.0');
  });

  it('should get the correct patch bump', () => {
    const nextVersion = getNextVersionFromLabels('1.0.0', config.user, changesWithPatch, false);

    expect(nextVersion).toBe('1.0.1');
  });

  it('should get the correct rc bump', () => {
    const nextVersion = getNextVersionFromLabels('1.2.3', config.user, changesWithMajor, true);

    expect(nextVersion).toBe('2.0.0-rc.0');
  });

  it('should get the correct bump from a rc', () => {
    const nextVersion = getNextVersionFromLabels('1.0.0-rc.2', config.user, changesWithPatch, false);

    expect(nextVersion).toBe('1.0.0');
  });

  it('should get the correct rc bump from a previous rc', () => {
    const nextVersion = getNextVersionFromLabels('1.0.1-rc.2', config.user, changesWithMinor, true);

    expect(nextVersion).toBe('1.1.0-rc.0');
  });

  it('should generate a changelog', () => {
    const forge = new GithubForge('', '');
    const changelog = getChangeLogSection('1.0.0', 'v1.0.0', config, changes, forge, true);

    expect(changelog).toMatchSnapshot();
  });

  const changelogFiles = [
    {
      name: 'should add new section',
      file: '__fixtures__/CHANGELOG_1.md',
      latestVersion: '1.0.0',
      nextVersion: '1.0.1',
    },
    {
      name: 'should update existing changelog section',
      file: '__fixtures__/CHANGELOG_2.md',
      latestVersion: '2.0.1',
      nextVersion: '2.0.3',
    },
    {
      name: 'should remove versions newer than the latest released version',
      // this happens if we bump the version in an existing release PR
      file: '__fixtures__/CHANGELOG_3.md',
      latestVersion: '2.0.1',
      nextVersion: '2.0.4',
    },
    {
      name: 'should remove previous prerelease versions as soon as the full release is added',
      file: '__fixtures__/CHANGELOG_4.md',
      latestVersion: '2.0.1-pre.1',
      nextVersion: '2.0.1',
    },
  ];
  it.each(changelogFiles)('$name', async ({ file, nextVersion, latestVersion }) => {
    const oldChangelog = await fs.readFile(path.join(__dirname, file), 'utf8');

    const forge = new GithubForge('', '');
    const newSection = getChangeLogSection(nextVersion, nextVersion, config, changes, forge, true);
    const changelog = updateChangelogSection(latestVersion, nextVersion, oldChangelog, newSection);

    expect(changelog).toMatchSnapshot();
  });

  it('should be able to extract the release version from a commit message', () => {
    const tests = [
      {
        commitMessage: 'ðŸŽ‰ Release 1.2.3',
        expectedVersion: '1.2.3',
      },
      {
        commitMessage: 'ðŸŽ‰ Release 1.2.3-rc.0 [skip ci]',
        expectedVersion: '1.2.3-rc.0',
      },
      {
        commitMessage: 'chore(release): 1.0.0-rc.0 [skip ci]',
        expectedVersion: '1.0.0-rc.0',
      },
    ];

    tests.forEach(({ commitMessage, expectedVersion }) => {
      expect(extractVersionFromCommitMessage(commitMessage)).toBe(expectedVersion);
    });
  });
});
