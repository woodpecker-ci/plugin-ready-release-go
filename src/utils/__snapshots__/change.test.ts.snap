// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`change > 'should add new section' 1`] = `
"# Changelog

## [1.0.1](https://github.com/woodpecker-ci/woodpecker/releases/tag/1.0.1) - 2000-02-01

### ❤️ Thanks to all contributors! ❤️

@John Doe, @Alice Wonderland

### 💥 Breaking changes

- Break everything [[#1339](https://github.com/woodpecker-ci/woodpecker/pull/1339)]

### ✨ Features

- Add new feature [[#1338](https://github.com/woodpecker-ci/woodpecker/pull/1338)]

### 🐛 Bug Fixes

- Fix random UI bug [[#1337](https://github.com/woodpecker-ci/woodpecker/pull/1337)]
- Fix another bug ([abc](https://github.com/woodpecker-ci/woodpecker/commit/abc))

### 📈 Enhancement

- Update README ([def](https://github.com/woodpecker-ci/woodpecker/commit/def))

### Misc

- Update README ([def](https://github.com/woodpecker-ci/woodpecker/commit/def))

## [1.0.0](https://github.com/woodpecker-ci/woodpecker/releases/tag/1.0.0) - 2023-07-11

### ✨ Features

- Add new feature [#1338](https://github.com/woodpecker-ci/woodpecker/pull/1338) ([456](https://github.com/woodpecker-ci/woodpecker/commit/456))

## [0.0.2](https://github.com/woodpecker-ci/woodpecker/releases/tag/0.0.1) - 2020-01-01

### 🐛 Bug Fixes

- Fix random UI bug [#1337](https://github.com/woodpecker-ci/woodpecker/pull/1337) ([123](https://github.com/woodpecker-ci/woodpecker/commit/123))
- Fix another bug [#43](https://github.com/woodpecker-ci/woodpecker/pull/43) ([456](https://github.com/woodpecker-ci/woodpecker/commit/456))

## [0.0.1](https://github.com/woodpecker-ci/woodpecker/releases/tag/0.0.1) - 2020-01-01

### ✨ Features

- Add new feature (#1338) @Alice Wonderland
"
`;

exports[`change > 'should remove versions newer than the latest released version' 1`] = `
"# Changelog

## [2.0.4](https://github.com/woodpecker-ci/woodpecker/releases/tag/2.0.4) - 2000-02-01

### ❤️ Thanks to all contributors! ❤️

@John Doe, @Alice Wonderland

### 💥 Breaking changes

- Break everything [[#1339](https://github.com/woodpecker-ci/woodpecker/pull/1339)]

### ✨ Features

- Add new feature [[#1338](https://github.com/woodpecker-ci/woodpecker/pull/1338)]

### 🐛 Bug Fixes

- Fix random UI bug [[#1337](https://github.com/woodpecker-ci/woodpecker/pull/1337)]
- Fix another bug ([abc](https://github.com/woodpecker-ci/woodpecker/commit/abc))

### 📈 Enhancement

- Update README ([def](https://github.com/woodpecker-ci/woodpecker/commit/def))

### Misc

- Update README ([def](https://github.com/woodpecker-ci/woodpecker/commit/def))

## [v2.0.1](https://github.com/woodpecker-ci/plugin-git/releases/tag/v2.0.1) - 2022-12-21

- BUGFIXES
  - Fix dockerfile to build correct go binary (#54)
  - Do not set GIT_TERMINAL_PROMPT=0 for git (#52)

## [v2.0.0](https://github.com/woodpecker-ci/plugin-git/releases/tag/v2.0.0) - 2022-11-14

- FEATURES
  - Partial clone by default (#48)
- BUGFIXES
  - Make home var settable (#47)
- ENHANCEMENTS
  - Publish semver images (#50)
"
`;

exports[`change > 'should update existing changelog section' 1`] = `
"# Changelog

## [2.0.3](https://github.com/woodpecker-ci/woodpecker/releases/tag/2.0.3) - 2000-02-01

### ❤️ Thanks to all contributors! ❤️

@John Doe, @Alice Wonderland

### 💥 Breaking changes

- Break everything [[#1339](https://github.com/woodpecker-ci/woodpecker/pull/1339)]

### ✨ Features

- Add new feature [[#1338](https://github.com/woodpecker-ci/woodpecker/pull/1338)]

### 🐛 Bug Fixes

- Fix random UI bug [[#1337](https://github.com/woodpecker-ci/woodpecker/pull/1337)]
- Fix another bug ([abc](https://github.com/woodpecker-ci/woodpecker/commit/abc))

### 📈 Enhancement

- Update README ([def](https://github.com/woodpecker-ci/woodpecker/commit/def))

### Misc

- Update README ([def](https://github.com/woodpecker-ci/woodpecker/commit/def))

## [v2.0.1](https://github.com/woodpecker-ci/plugin-git/releases/tag/v2.0.1) - 2022-12-21

- BUGFIXES
  - Fix dockerfile to build correct go binary (#54)
  - Do not set GIT_TERMINAL_PROMPT=0 for git (#52)

## [v2.0.0](https://github.com/woodpecker-ci/plugin-git/releases/tag/v2.0.0) - 2022-11-14

- FEATURES
  - Partial clone by default (#48)
- BUGFIXES
  - Make home var settable (#47)
- ENHANCEMENTS
  - Publish semver images (#50)

## [v1.6.1](https://github.com/woodpecker-ci/plugin-git/releases/tag/v1.6.1) - 2022-11-06

- BUGFIXES
  - Explicite set and check for home dir (#46)

## [v1.6.0](https://github.com/woodpecker-ci/plugin-git/releases/tag/v1.6.0) - 2022-10-13

- BUGFIXES
  - Handle git-lfs separately (#40)
- ENHANCEMENTS
  - if no branch info is set, fallback to default repo branch (#41)

## [v1.5.0](https://github.com/woodpecker-ci/plugin-git/releases/tag/v1.5.0) - 2022-10-06

- ENHANCEMENTS
  - Release binarys (#37)
  - Use ref to checkout if no commit sha is set (#36)
  - Fix tests (#35)
- MISC
  - Update urfave/cli to v2.17.1 (#38)
  - Use built-in log instead of logrus (#34)

## [v1.4.0](https://github.com/woodpecker-ci/plugin-git/releases/tag/v1.4.0) - 2022-08-30

- ENHANCEMENTS
  - Auto enable tags clone if it's ci event is 'tag' (#30)
  - Support more architectures (#29)

## [v1.3.0](https://github.com/woodpecker-ci/plugin-git/releases/tag/v1.3.0) - 2022-08-15

- FEATURES
  - Add option to Change branch name for checkout (#28)

## [v1.2.0](https://github.com/woodpecker-ci/plugin-git/releases/tag/v1.2.0) - 2022-05-25

- FEATURES
  - Add git-lfs (#21)
  - Custom ssl certs for git (#19)
- ENHANCEMENTS
  - Add an \`lfs\` setting which lets you disable Git LFS (#24)
- DOCUMENTATION
  - Add docs page (#23)

## [v1.1.2](https://github.com/woodpecker-ci/plugin-git/releases/tag/v1.1.2) - 2022-01-30

- BUGFIXES
  - Fix empty login/password in netrc (#20)

## [v1.1.1](https://github.com/woodpecker-ci/plugin-git/releases/tag/v1.1.1) - 2021-12-23

- BUGFIXES
  - Fix version info (#13)

## [v1.1.0](https://github.com/woodpecker-ci/plugin-git/releases/tag/v1.1.0) - 2021-12-18

- FEATURES
  - Add ppc64le support (#8)
- BUGFIXES
  - Regognize \\"CI_*\\" EnvVars (#6)
- ENHANCEMENTS
  - Multiarch build (#8)
- MISC
  - Upgrade urfave/cli to v2 (#5)
"
`;

exports[`change > should generate a changelog 1`] = `
"## [1.0.0](https://github.com/woodpecker-ci/woodpecker/releases/tag/1.0.0) - 2000-02-01

### ❤️ Thanks to all contributors! ❤️

@John Doe, @Alice Wonderland

### 💥 Breaking changes

- Break everything [[#1339](https://github.com/woodpecker-ci/woodpecker/pull/1339)]

### ✨ Features

- Add new feature [[#1338](https://github.com/woodpecker-ci/woodpecker/pull/1338)]

### 🐛 Bug Fixes

- Fix random UI bug [[#1337](https://github.com/woodpecker-ci/woodpecker/pull/1337)]
- Fix another bug ([abc](https://github.com/woodpecker-ci/woodpecker/commit/abc))

### 📈 Enhancement

- Update README ([def](https://github.com/woodpecker-ci/woodpecker/commit/def))

### Misc

- Update README ([def](https://github.com/woodpecker-ci/woodpecker/commit/def))"
`;
